src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,398 @@
+
+                    if (_integerDecoder.BeginDecode((byte)(b & ~HuffmanMask), StringLengthPrefix))
+                    {
+                        OnStringLength(_integerDecoder.Value, nextState: State.HeaderName);
+                    }
+                    else
+                    {
+                        _state = State.HeaderNameLengthContinue;
+                    }
+
+                    break;
+                case State.HeaderNameLengthContinue:
+                    if (_integerDecoder.Decode(b))
+                    {
+                        OnStringLength(_integerDecoder.Value, nextState: State.HeaderName);
+                    }
+
+                    break;
+                case State.HeaderName:
+                    _stringOctets[_stringIndex++] = b;
+
+                    if (_stringIndex == _stringLength)
+                    {
+                        OnString(nextState: State.HeaderValueLength);
+                    }
+
+                    break;
+                case State.HeaderValueLength:
+                    _huffman = (b & HuffmanMask) != 0;
+
+                    if (_integerDecoder.BeginDecode((byte)(b & ~HuffmanMask), StringLengthPrefix))
+                    {
+                        OnStringLength(_integerDecoder.Value, nextState: State.HeaderValue);
+                    }
+                    else
+                    {
+                        _state = State.HeaderValueLengthContinue;
+                    }
+
+                    break;
+                case State.HeaderValueLengthContinue:
+                    if (_integerDecoder.Decode(b))
+                    {
+                        OnStringLength(_integerDecoder.Value, nextState: State.HeaderValue);
+                    }
+
+                    break;
+                case State.HeaderValue:
+                    _stringOctets[_stringIndex++] = b;
+
+                    if (_stringIndex == _stringLength)
+                    {
+                        OnString(nextState: State.Ready);
+
+                        var headerNameSpan = new Span<byte>(_headerName, 0, _headerNameLength);
+                        var headerValueSpan = new Span<byte>(_headerValueOctets, 0, _headerValueLength);
+
+                        onHeader(headerNameSpan, headerValueSpan);
+
+                        if (_index)
+                        {
+                            _dynamicTable.Insert(headerNameSpan, headerValueSpan);
+                        }
+                    }
+
+                    break;
+                case State.DynamicTableSizeUpdate:
+                    if (_integerDecoder.Decode(b))
+                    {
+                        if (_integerDecoder.Value > _maxDynamicTableSize)
+                        {
+                            throw new HPackDecodingException(
+                                $"CoreStrings.FormatHPackErrorDynamicTableSizeUpdateTooLarge({_integerDecoder.Value}, {_maxDynamicTableSize})");
```

Comment:
Before checking in, it'd be good to resource-ify all strings that need it.

Reply:
Most likely, I will just remove the strings. We should tackle the general error reporting problem first.